base/compiler/ssair/slot2ssa.jl

```diff
@@ -211,15 +211,24 @@ struct DelayedTyp
 end
 
 # maybe use expr_type?
-function typ_for_val(@nospecialize(val), ci::CodeInfo)
-    isa(val, Expr) && return val.typ
-    isa(val, GlobalRef) && return abstract_eval_global(val.mod, val.name)
-    isa(val, SSAValue) && return ci.ssavaluetypes[val.id]
-    isa(val, Argument) && return ci.slottypes[val.n]
-    isa(val, NewSSAValue) && return DelayedTyp(val)
-    isa(val, QuoteNode) && return Const(val.value)
-    isa(val, Union{Symbol, PiNode, PhiNode, SlotNumber, TypedSlot}) && error("unexpected val type")
-    return Const(val)
+function typ_for_val(@nospecialize(x), ci::CodeInfo, spvals::SimpleVector, idx::Int)
+    if isa(x, Expr)
+        if x.head === :static_parameter
+            return sparam_type(spvals[x.args[1]])
+        elseif x.head === :boundscheck
+            return Bool
```

Comment:
why are these cases necessary? It seems like it would be very bad if `ci.ssavaluetypes[idx]` didn't also have the right answer

Reply:
These can occur as arguments to calls.